src/cli.rs

```diff
@@ -695,6 +707,99 @@ impl Opts {
         Ok(())
     }
 
+    async fn import_flake(&self, o: &ImportFlakeOpts) -> Result<()> {
+        let mut pins = self.read_pins()?;
+
+        let flake: serde_json::Value =
+            serde_json::from_reader(std::fs::File::open(&o.path).context(anyhow::format_err!(
+                "Could not open flake.lock at '{}'",
+                o.path.canonicalize().unwrap_or_else(|_| o.path.clone()).display()
+            ))?)
+            .context("Nix lock file is not a valid JSON dict")?;
+        log::info!("Note that all the imported entries will be updated so they won't necessarily point to the same commits as before!");
+
+        let nodes: &serde_json::Map<String, serde_json::Value> = flake
+            .get("nodes")
+            .context("flake.lock missing key `nodes`")?
+            .as_object()
+            .context("`nodes` key does not contain an object")?;
+
+        let root_name = flake
+            .get("root")
+            .context("missing `root` key")?
+            .as_str()
+            .context("")?;
+        let root = nodes
+            .get(root_name)
+            .context("flake.lock missing key `root`")?
+            .get("inputs")
+            .context("`root` key missing `inputs` key")?
+            .as_object()
+            .context("`root` key is not an object")?;
+
+        let inputs: BTreeMap<String, String> = root
+            .into_iter()
+            .map(|(key, value)| Some((key.to_string(), value.as_str()?.to_string())))
+            .collect::<Option<_>>()
+            .context("")?;
+
+        async fn import(
+            name: &str,
+            npins: &mut NixPins,
+            nodes: &serde_json::Map<String, serde_json::Value>,
+        ) -> Result<()> {
+            let pin = nodes
+                .get(name)
+                .ok_or_else(|| anyhow::format_err!("Pin '{}' not found in flake.lock", name))?;
+            anyhow::ensure!(
+                !npins.pins.contains_key(name),
+                "Pin '{}' exists in both files, this is a collision. Please delete the entry in one of the files.",
+                name
+            );
+
+            let pin: flake::FlakePin = serde_json::from_value(pin.clone())?;
+            // .context("Pin is either invalid, or we don't support it")?;
+
+            if pin.is_indirect() {
+                log::info!("skipping indirect input {}", name);
```

Comment:
I'm not sure I understand you here — do you mean that we could support importing indirect inputs as well?

Indirect inputs may be a different name for an input already used elsewhere, but in general they are (I think) any kind of transitive dependency, i.e. may point to another version of nixpkgs used by a dependency or something like that.

Reply:
My thought was that "indirect" dependencies are all those who are declared as `inputs.<foo>.inputs.nixpkgs.follows = "other";` (or whatever the right structure is).

I don't think we currently have a way to compose a Flake with all its transitive inputs. Perhaps that is something to consider for the future. Given how "simple" the flake composition expression is: https://github.com/NixOS/nix/blob/b4950404ba7c8f14f5892bd1327b035590c2f7bc/src/libexpr/flake/call-flake.nix

For now, I'd not block on this.

Reply:
> My thought was that "indirect" dependencies are all those who are declared as inputs.<foo>.inputs.nixpkgs.follows = "other"; (or whatever the right structure is).

Not quite — these are indirect inputs, yes, but also all transitive dependencies (even without a `follows`) are also indirect inputs; a flake resolves all its transitive dependencies in its own lock file (which npins does not — if I have one npins-managed repository depending on another also managed with npins or a similar tool, I do not know anything about transitive dependencies if the latter has not decided to expose them).

> Perhaps that is something to consider for the future. Given how "simple" the flake composition expression is

It might be cool to have, yes (indeed I have a similar function in my current nixos config so that I can still used some dependencies which only have a `flake.nix` as entrypoint).

But as I said in the original PR message, I'm not sure how to automate that as part of an `import-flake` command — at the very least, it'd need some major changes to the `npins/default.nix` file, and possible also the structure of `sources.json` to model indirect inputs (& keeping track of transitive dependency relations), so I've not attempted that for now; doing it manually seemed "good enough", at least for my own migration away from flakes.